File: android/app/src/main/java/dev/seedseeker/app/model/Presets.kt
// SPDX-License-Identifier: GPL-3.0-or-later
package dev.seedseeker.app.model

import dev.seedseeker.app.catalog.ItemCatalog
import java.util.UUID

data class PresetQuery(
    val requirements: List<ItemRequirement>,
    val maximumDepth: Int = 24,
    val requireBlacksmith: Boolean = false,
    val excludeBlacksmithRewards: Boolean = false,
    val wandmakerQuest: WandmakerQuest? = null,
    val challenges: Int = 0,
    val autoApplyTrinket: Boolean = true,
)

data class QueryPreset(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val query: PresetQuery,
    val isBuiltIn: Boolean = false,
)

object BuiltInPresets {
    /**
     * The floor limit the vault presets carry: floor 19 is the last floor the
     * Imp — and so the vault holding its levelled prizes — can appear on, so a
     * deeper scan only costs time.
     */
    private const val VAULT_FLOOR_LIMIT = 19

    val staff21 = QueryPreset(
        id = "staff-21",
        name = "+21 Staff",
        isBuiltIn = true,
        query = PresetQuery(
            requirements = listOf(
                ItemRequirement(1, null, 3, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.EXACT, identityGroup = 1),
                ItemRequirement(2, null, 0, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.ANY, identityGroup = 1),
                ItemRequirement(3, null, 0, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.ANY, identityGroup = 1),
                ItemRequirement(4, null, 1, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.AT_LEAST),
            ),
        ),
    )

    /**
     * The +21 stack anchored one level higher, on the +4 wand v4.0.0's Imp
     * vault lays out among its prizes.
     */
    val staff22 = QueryPreset(
        id = "staff-22",
        name = "+22 Staff",
        isBuiltIn = true,
        query = PresetQuery(
            requirements = listOf(
                ItemRequirement(1, null, 4, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.EXACT, identityGroup = 1),
                ItemRequirement(2, null, 0, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.ANY, identityGroup = 1),
                ItemRequirement(3, null, 0, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.ANY, identityGroup = 1),
                ItemRequirement(4, null, 1, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.AT_LEAST),
            ),
            maximumDepth = VAULT_FLOOR_LIMIT,
        ),
    )

    val wandBonanza = QueryPreset(
        id = "wand-bonanza",
        name = "Wand Bonanza",
        isBuiltIn = true,
        query = PresetQuery(
            requirements = listOf(
                ItemRequirement(1, null, 3, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.EXACT),
                ItemRequirement(2, null, 2, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.EXACT, maximumDepth = 4),
                ItemRequirement(3, null, 2, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.EXACT, maximumDepth = 4),
                ItemRequirement(4, null, 2, kind = ItemKind.WAND, upgradeMatch = UpgradeMatch.EXACT),
            ),
        ),
    )

    val ringOfWealth21 = QueryPreset(
        id = "ring-of-wealth-21",
        name = "+21 Ring of Wealth",
        isBuiltIn = true,
        query = PresetQuery(
            requirements = listOf(
                ItemRequirement(
                    1,
                    ItemCatalog.findById("ring_wealth"),
                    4,
                    kind = ItemKind.RING,
                    upgradeMatch = UpgradeMatch.EXACT,
                    source = ScoutItemSource.IMP_REWARD,
                ),
                ItemRequirement(
                    2,
                    ItemCatalog.findById("ring_wealth"),
                    2,
                    kind = ItemKind.RING,
                    upgradeMatch = UpgradeMatch.EXACT,
                ),
                ItemRequirement(
                    3,
                    ItemCatalog.findById("ring_wealth"),
                    0,
                    kind = ItemKind.RING,
                    upgradeMatch = UpgradeMatch.ANY,
                ),
            ),
        ),
    )

    /**
     * A tier-4 weapon at the +5 only the vault reaches, with two more of the
     * same weapon to pour into it.
     */
    val tier4Weapon26 = QueryPreset(
        id = "tier-4-weapon-26",
        name = "+26 Tier 4 Weapon",
        isBuiltIn = true,
        query = PresetQuery(
            requirements = listOf(
                ItemRequirement(
                    1,
                    null,
                    5,
                    kind = ItemKind.WEAPON,
                    tier = 4,
                    tierMatch = TierMatch.EXACT,
                    upgradeMatch = UpgradeMatch.EXACT,
                    identityGroup = 1,
                ),
                ItemRequirement(2, null, 0, kind = ItemKind.WEAPON, upgradeMatch = UpgradeMatch.ANY, identityGroup = 1),
                ItemRequirement(3, null, 0, kind = ItemKind.WEAPON, upgradeMatch = UpgradeMatch.ANY, identityGroup = 1),
            ),
            maximumDepth = VAULT_FLOOR_LIMIT,
        ),
    )

    val all = listOf(staff21, staff22, wandBonanza, ringOfWealth21, tier4Weapon26)
}

File: linux/src/presets.rs
// SPDX-License-Identifier: GPL-3.0-or-later

//! Presets bundled with every installation.

use shpd_seedfinder_core::catalog::{ItemId, ItemKind};
use shpd_seedfinder_core::model::ItemSource;
use shpd_seedfinder_core::query::UpgradeRequirement;

use crate::state::{AppState, UiRequirement};

/// One read-only query shipped with the application.
#[derive(Clone, Debug)]
pub struct BuiltInPreset {
    pub name: &'static str,
    pub state: AppState,
}

/// Returns the protected presets in presentation order.
#[must_use]
pub fn built_in() -> [BuiltInPreset; 3] {
    [staff_21(), wand_bonanza(), ring_of_wealth_21()]
}

fn wand_bonanza() -> BuiltInPreset {
    let mut state = AppState::default();
    for (upgrade, max_depth) in [
        (UpgradeRequirement::Exact(3), None),
        (UpgradeRequirement::Exact(2), Some(4)),
        (UpgradeRequirement::Exact(2), Some(4)),
        (UpgradeRequirement::Exact(2), None),
    ] {
        let key = state.claim_key();
        state.requirements.push(UiRequirement {
            key,
            kind: ItemKind::Wand,
            upgrade,
            max_depth,
            ..UiRequirement::new(key)
        });
    }
    BuiltInPreset {
        name: "Wand Bonanza",
        state,
    }
}

fn staff_21() -> BuiltInPreset {
    let mut state = AppState::default();
    for (upgrade, identity_group) in [
        (UpgradeRequirement::Exact(3), Some(1)),
        (UpgradeRequirement::Any, Some(1)),
        (UpgradeRequirement::Any, Some(1)),
        (UpgradeRequirement::AtLeast(1), None),
    ] {
        let key = state.claim_key();
        state.requirements.push(UiRequirement {
            key,
            kind: ItemKind::Wand,
            upgrade,
            identity_group,
            ..UiRequirement::new(key)
        });
    }
    BuiltInPreset {
        name: "+21 Staff",
        state,
    }
}

fn ring_of_wealth_21() -> BuiltInPreset {
    let mut state = AppState::default();
    for (upgrade, source) in [
        (UpgradeRequirement::Exact(4), Some(ItemSource::ImpReward)),
        (UpgradeRequirement::Exact(2), None),
        (UpgradeRequirement::Any, None),
    ] {
        let key = state.claim_key();
        state.requirements.push(UiRequirement {
            key,
            kind: ItemKind::Ring,
            item: Some(ItemId::RingWealth),
            upgrade,
            source,
            ..UiRequirement::new(key)
        });
    }
    BuiltInPreset {
        name: "+21 Ring of Wealth",
        state,
    }
}

#[cfg(test)]
mod tests {
    use shpd_seedfinder_core::catalog::{ItemId, ItemKind};
    use shpd_seedfinder_core::model::ItemSource;
    use shpd_seedfinder_core::query::UpgradeRequirement;

    use super::built_in;

    #[test]
    fn staff_matches_requested_requirements() {
        let [staff, _, _] = built_in();
        assert_eq!(staff.name, "+21 Staff");
        assert_eq!(staff.state.requirements.len(), 4);
        assert!(
            staff
                .state
                .requirements
                .iter()
                .all(|requirement| requirement.kind == ItemKind::Wand)
        );
        assert_eq!(
            staff
                .state
                .requirements
                .iter()
                .map(|requirement| requirement.upgrade)
                .collect::<Vec<_>>(),
            [
                UpgradeRequirement::Exact(3),
                UpgradeRequirement::Any,
                UpgradeRequirement::Any,
                UpgradeRequirement::AtLeast(1),
            ]
        );
        assert_eq!(
            staff
                .state
                .requirements
                .iter()
                .map(|requirement| requirement.identity_group)
                .collect::<Vec<_>>(),
            [Some(1), Some(1), Some(1), None]
        );
    }

    #[test]
    fn wand_bonanza_matches_requested_requirements() {
        let [_, preset, _] = built_in();
        assert_eq!(preset.name, "Wand Bonanza");
        assert!(
            preset
                .state
                .requirements
                .iter()
                .all(|requirement| requirement.kind == ItemKind::Wand && requirement.item.is_none())
        );
        assert_eq!(
            preset
                .state
                .requirements
                .iter()
                .map(|requirement| requirement.upgrade)
                .collect::<Vec<_>>(),
            [
                UpgradeRequirement::Exact(3),
                UpgradeRequirement::Exact(2),
                UpgradeRequirement::Exact(2),
                UpgradeRequirement::Exact(2),
            ]
        );
        assert_eq!(
            preset
                .state
                .requirements
                .iter()
                .map(|requirement| requirement.max_depth)
                .collect::<Vec<_>>(),
            [None, Some(4), Some(4), None]
        );
        assert!(
            preset
                .state
                .requirements
                .iter()
                .all(|requirement| requirement.identity_group.is_none())
        );
    }

    #[test]
    fn ring_of_wealth_matches_requested_requirements() {
        let [_, _, ring] = built_in();
        assert_eq!(ring.name, "+21 Ring of Wealth");
        assert!(
            ring.state
                .requirements
                .iter()
                .all(|requirement| requirement.item == Some(ItemId::RingWealth))
        );
        assert_eq!(
            ring.state.requirements[0].upgrade,
            UpgradeRequirement::Exact(4)
        );
        assert_eq!(
            ring.state.requirements[0].source,
            Some(ItemSource::ImpReward)
        );
        assert_eq!(
            ring.state
                .requirements
                .iter()
                .map(|requirement| requirement.max_depth)
                .collect::<Vec<_>>(),
            [None, None, None]
        );
        assert_eq!(
            ring.state.requirements[1].upgrade,
            UpgradeRequirement::Exact(2)
        );
        assert_eq!(ring.state.requirements[2].upgrade, UpgradeRequirement::Any);
    }
}
